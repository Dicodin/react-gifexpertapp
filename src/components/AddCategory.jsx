// Iniciando todo con el snnipet "rafc"
import React, { useState } from 'react';
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {

    // El estado se inicia con "inputValue = ''"
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        // Para igualar nuestro estado a lo que escribamos en el input
        setInputValue(e.target.value);

    }

    const handleSubmit = (e) => {
        e.preventDefault();

        if (inputValue.trim().length > 2) {
            // De esta forma si podemos mutar el estado del arreglo que viene "dentro" de la función de nuestro prop "setCategories"
            setCategories(categ => [inputValue, ...categ]);
            // Esta es la función que llamamos de nuestro estado para "limpiar" nuestro input
            setInputValue('');
        }

    }

    return (
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                placeholder='Buscar gifs'
                value={inputValue}
                onChange={handleInputChange}
            ></input>
        </form>
    )
}

// Validamos que nuestro prop "setCategories" sea del tipo correspondiente "func" y que sea requerido
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
