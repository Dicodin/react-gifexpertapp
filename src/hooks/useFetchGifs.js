// Todos los "hooks" inician on "use" en su nombre
// Este es un "custom hook"
// Los "custom hooks" son como "Functional Cmponents"
import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs";

export const useFecthGifs = (category) => {

    const [state, setstate] = useState({
        data: [],
        loading: true
    });

    // "useEffect" para indicar que quiero que la función "getGifs" se ejecute una única vez si el arreglo como segundo argumento esta vacío
    // Enviamos "category" como parámetro en el arreglo por si deseamos que el componenete se vuelva a jecutar por medio del "useEffect"
    // Los "useEffect" no pueden ser "async"
    useEffect(() => {
        getGifs(category)
            .then(imgs => {
                setstate({
                    data: imgs,
                    loading: false
                });
            });
    }, [category]);

    return state; // {data: [], loading: true}
}