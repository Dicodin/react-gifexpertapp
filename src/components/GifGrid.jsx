import React from 'react'
import { useFecthGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    // Uso mi "custom hook" "useFecthGifs"
    const { data: images, loading } = useFecthGifs(category);

    return (
        <>
            <h3 className="animate__animated animate__fadeIn">{category}</h3>

            {/* Para abreviar la condici√≥n "if" con un solo resultado */}
            {loading && <p className="animate__animated animate__flash">Loading</p>}

            <div className="card-grid">
                {
                    images.map(img => (
                        <GifGridItem
                            key={img.id} // Cada elemento debe tener su key
                            {...img} // Usamos el operador spread para enviar todas las props
                        ></GifGridItem>
                    ))
                }
            </div>
        </>
    )
}