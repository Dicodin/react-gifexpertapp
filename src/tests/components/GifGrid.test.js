import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFecthGifs } from "../../hooks/useFetchGifs";

// Para simular cualquier llamada al archivo "useFetchGifs" para posteriormente controlar lo que va a responder
jest.mock("../../hooks/useFetchGifs");

describe('Pruebas sobre <GifGrid></GifGrid>', () => {

    const category = 'Tao';

    test('debe de mostrar correctamente el componente ', () => {

        // Para simular que realizamos una petición, pero en este caso retorna el estado inicial
        useFecthGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrid category={category}></GifGrid>);
        expect(wrapper).toMatchSnapshot();

    });

    test('debe de mostrar items cuando se cargan imagenes con useFecthGifs ', () => {

        // Debido a que "getGifs" esta difinido de esta forma
        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/cualquier/cosa.jpg',
            title: 'Cualquier cosa'
        }];

        // Simulamos que ya obtuvimos un resultado, por lo tanto el "loading" terimnará en "false"
        useFecthGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow(<GifGrid category={category}></GifGrid>);

        // expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);

    });

});
