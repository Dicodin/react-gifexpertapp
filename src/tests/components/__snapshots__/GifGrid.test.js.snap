// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <GifGrid></GifGrid> debe de mostrar correctamente el componente  1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Tao
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
